### ISSUE TITLE
ArgumentError at /node/version

### ISSUE THREAD

**Issue by murrant** [ytti/oxidized-web#38]
I'm using groups and I get this error when clicking on versions.

```
ArgumentError at /node/version
wrong number of arguments (0 for 1+)
file: versions.haml
location: count
line: 41
```

**Comment by ytti** [comment]
On 30 November 2015 at 18:49, Tony Murray <notifications@github.com> wrote:

> line: 41
> BACKTRACE
> (expand)
> JUMP TO: GET POST COOKIES ENV
>

​Do you have the backtrace?
​



-- 
  ++ytti

**Comment by murrant** [comment]
Sorry, got cut out of my silly attempt to copy and paste:

/usr/local/lib64/ruby/gems/2.0.0/gems/oxidized-web-0.5.2/lib/oxidized/web/views/versions.haml in count
41.      - nb = @data.count + 1

/usr/local/lib64/ruby/gems/2.0.0/gems/oxidized-web-0.5.2/lib/oxidized/web/views/versions.haml in block in singleton class
41.      - nb = @data.count + 1

/usr/local/lib64/ruby/gems/2.0.0/gems/tilt-2.0.1/lib/tilt/template.rb in call
155.      method.bind(scope).call(locals, &block)

/usr/local/lib64/ruby/gems/2.0.0/gems/tilt-2.0.1/lib/tilt/template.rb in evaluate
155.      method.bind(scope).call(locals, &block)

/usr/local/lib64/ruby/gems/2.0.0/gems/tilt-2.0.1/lib/tilt/haml.rb in evaluate
17.        super

/usr/local/lib64/ruby/gems/2.0.0/gems/tilt-2.0.1/lib/tilt/template.rb in render
96.      evaluate(scope, locals || {}, &block)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in render
822.        output          = template.render(scope, locals, &block)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in haml
676.      render(:haml, template, options, locals, &block)

/usr/local/lib64/ruby/gems/2.0.0/gems/oxidized-web-0.5.2/lib/oxidized/web/webapp.rb in out
229.          haml template, :layout => true

/usr/local/lib64/ruby/gems/2.0.0/gems/oxidized-web-0.5.2/lib/oxidized/web/webapp.rb in block in <class:WebApp>
163.        out :versions

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call
1610.          proc { |a,p| unbound_method.bind(a).call }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in compile!
1610.          proc { |a,p| unbound_method.bind(a).call }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in []
974.            route_eval { block[*args] }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block (3 levels) in route!
974.            route_eval { block[*args] }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in route_eval
993.      throw :halt, yield

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block (2 levels) in route!
974.            route_eval { block[*args] }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in process_route
1014.        block ? block[self, values] : yield(self, values)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in catch
1012.      catch(:pass) do

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in process_route
1012.      catch(:pass) do

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in route!
972.          returned_pass_block = process_route(pattern, keys, conditions) do |*args|

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in each
971.        routes.each do |pattern, keys, conditions, block|

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in route!
971.        routes.each do |pattern, keys, conditions, block|

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in dispatch!
1084.        route!

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in invoke
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in catch
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in invoke
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in dispatch!
1081.      invoke do

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in call!
906.      invoke { dispatch! }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in invoke
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in catch
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in invoke
1066.      res = catch(:halt) { yield }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call!
906.      invoke { dispatch! }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call
894.      dup.call!(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb in call
18.        status, headers, body = @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/path_traversal.rb in call
16.        app.call env

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/json_csrf.rb in call
18.        status, headers, body = app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb in call
49.        result or app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb in call
49.        result or app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-protection-1.5.3/lib/rack/protection/frame_options.rb in call
31.        status, headers, body        = @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/nulllogger.rb in call
9.      @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/head.rb in call
13.    status, headers, body = @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/show_exceptions.rb in call
21.      @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call
181.      result, callback = app.call(env), env['async.callback']

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call
2021.      @stack.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in block in call
1486.        synchronize { prototype.call(env) }

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in synchronize
1795.          yield

/usr/local/lib64/ruby/gems/2.0.0/gems/sinatra-1.4.6/lib/sinatra/base.rb in call
1486.        synchronize { prototype.call(env) }

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/urlmap.rb in block in call
66.        return app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/urlmap.rb in each
50.      @mapping.each do |host, location, match, app|

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/urlmap.rb in call
50.      @mapping.each do |host, location, match, app|

/usr/local/lib64/ruby/gems/2.0.0/gems/rack-1.6.4/lib/rack/builder.rb in call
153.      to_app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/puma-2.15.3/lib/puma/server.rb in handle_request
541.          status, headers, res_body = @app.call(env)

/usr/local/lib64/ruby/gems/2.0.0/gems/puma-2.15.3/lib/puma/server.rb in process_client
388.          case handle_request(client, buffer)

/usr/local/lib64/ruby/gems/2.0.0/gems/puma-2.15.3/lib/puma/server.rb in block in run
270.            process_client client, buffer

/usr/local/lib64/ruby/gems/2.0.0/gems/puma-2.15.3/lib/puma/thread_pool.rb in call
106.            block.call(work, *extra)

/usr/local/lib64/ruby/gems/2.0.0/gems/puma-2.15.3/lib/puma/thread_pool.rb in block in spawn_thread
106.            block.call(work, *extra)

**Comment by ytti** [comment]
Are you using 'file' output, instead of 'git'? We don't support 'version' for file. But the error message should be better.
In git version it now shouldn't puke, just shouldn't print anything. 

Not sure what we should do for 'file', as they don't really have concept of version. Perhaps timestamp of a file, but currently we always overwrite the file. So then we'd need to compare first if it changed and overwrite only if it didn't. 
Wouldn't be too hard, but unsure if very useful.

**Comment by murrant** [comment]
No, I'm using the git output.  https://fnpaste.com/AYJJ

**Comment by precompiled** [comment]
Can you please exclude me from your conversation, it has nothing to do with
me? Thanks

**Comment by murrant** [comment]
Not sure how you got on here precompiled.  You'll have to go to github issue and click the unsubscribe button. https://github.com/ytti/oxidized-web/issues/38

**Comment by precompiled** [comment]
Ok, I'm out, but if you reply to mail, I'll get message again.

**Comment by supertylerc** [comment]
@murrant, is this still an issue?  If it is, I'll set up a clean environment later this week using groups and see if I can replicate the issue.

**Comment by murrant** [comment]
Yes, it is still occurring.  FYI, I am using this branch of librenms: https://github.com/murrant/librenms/tree/oxidized_groups

**Comment by precompiled** [comment]
please exclude me from this conversation

**Comment by murrant** [comment]
@precompiled  we cannot, it is automated emails.  You removed yourself last time, but then you replied to the email and re-added yourself...  You will have to remove yourself again.  I'm sorry you got added initially, it was due to me pasting the large text in and it had an @ reply with your username.

**Comment by precompiled** [comment]
Then please fix this issue asap :)

**Comment by supertylerc** [comment]
@murrant, I've confirmed the issue.  I believe it is because the `GET` is using the following parameter:

```
node_full=cows/192.168.33.2
```

A forward slash isn't valid in a parameter (it's a path separator) and must be escaped (`%2F` is the URI escape character for `/`).

However, the better approach is probably to add a `group` parameter as specifying the group actually isn't mandatory.  So the path and parameters would look something like this:

```
/node/version?node_full=192.168.33.2&group=cows
```

@ytti, would that be okay?  I suspect this is broken for everyone, so introducing this change shouldn't break anything for anyone.

@murrant, would this solution work for you?

In either case, a fix should be quick and straightforward.

**Comment by supertylerc** [comment]
I lied.  Don't think that's the cause.  :)  Still digging.

**Comment by supertylerc** [comment]
@murrant:

As a temporary workaround, you can set the `groups` hash to an empty hash (either remove it or manually set `groups: {}`).  Oxidized will still set the correct group.  I'm looking into a permanent fix.

**Comment by supertylerc** [comment]
Found the line that causes the issue.  [See here](https://github.com/ytti/oxidized/blob/master/lib/oxidized/input/ssh.rb#L24).  Removing `vars(:ssh_port)` resolves the issue.  Inspecting it when setting groups reveals this:

```
#<Asetus::ConfigStruct:0x0000000259f090 @key_to_s=nil, @cfg={"to_i"=>#<Asetus::ConfigStruct:0x0000000259e320 @key_to_s=nil, @cfg={"to_str"=>#<Asetus::ConfigStruct:0x0000000258f5c8 @key_to_s=nil, @cfg={}>}>}>
```

Inspecting it when not using groups:

```
22
```

Still digging into a fix, but at least I know what's happening now.  :)

**Comment by supertylerc** [comment]
Commenting out [this line](https://github.com/ytti/oxidized/blob/master/lib/oxidized/config/vars.rb#L6) also resolves the issue.  I'm going to look into fixing that.

**Comment by supertylerc** [comment]
This will be fixed by ytti/oxidized#253.  Note that you'll need to manually build the gem and install it until a new gem is published.

**Comment by ytti** [comment]
Mr precompiled was added here, because the original ticket contained atprecompiled, I guarded the original ticket in backticks, so we likely are not spamming him anymore.
Apologies if he is still following.

**Comment by murrant** [comment]
I am still having this issue with Oxidized-0.10.0 and Oxidized-Web-0.6.1.

**Comment by supertylerc** [comment]
@murrant, I'm unable to reproduce. Can you please provide your full Oxidized config with indentation preserved? Also, please confirm if you experience this issue when using a CSV instead of a LibreNMS branch.

Thanks!

**Comment by murrant** [comment]
I have been using CSV, also LibreNMS has mainline group support now (different from that branch) You should also note, that I use single_repo.
```
output:
  default: git
  git:
    single_repo: true
```

**Comment by supertylerc** [comment]
@murrant, I can't properly reproduce this issue and begin investigation until you provide your full Oxidized configuration.  Please also provide a sample of your CSV file.

Thank you!

**Comment by supertylerc** [comment]
@murrant a current stack trace will also help.

**Comment by murrant** [comment]
Config: https://ptpb.pw/u-at
CSV: https://ptpb.pw/y9kY
Log with backtace: https://ptpb.pw/sBX6

**Comment by murrant** [comment]
I ended up disabling single_repo.  I do like that it stored things in one repo, but much prefer a working system.

**Comment by doogieconsulting** [comment]
Hey @ytti  i am getting the same error when i enable single_repo. Do you need/want me to provide any config files etc. for you to use? I am using LibreNMS HTTP source and git as the output.

**Comment by ytti** [comment]
I've not really looked into this, but it looks like @supertylerc has done quite bit of work on it. So being lazy bugger, I hope he'll figure it out.

**Comment by murrant** [comment]
@ytti @supertylerc  It is pretty simple to reproduce, you just need the following config.

If you have a repository with the same name as the group (<group>.git) it will pull the configs from there (instead of default.git/<group>) instead of crashing.   So, it is a bug either way.

```
output:
  default: git
  git:
    single_repo: true
    repo: /var/lib/oxidized/devices.git
source:
  default: csv
  csv:
    file: /var/lib/oxidized/devices.csv
    delimiter: !ruby/regexp /:/
    map:
      name: 0
      model: 1
      group: 2
      username: 3
      password: 4
```

**Comment by supertylerc** [comment]
Thanks!

I'll look at this issue this weekend. Wish I could look at it sooner, but this is a very long and busy week at work.

**Comment by yesbox** [comment]
This is indeed an issue with single_repo. It appears the oxidized-web still tries to access the git repo as if single_repo was false, as if repos with the group names exist (easily reproducible by changing single_repo from false to true), those will be presented rather than the diffs from the single config repo. When the group repos are removed [this error](https://github.com/ytti/oxidized-web/issues/38#issue-119532430) is shown.

**Comment by yesbox** [comment]
@supertylerc Are you looking into this? Would you need something else?

**Comment by supertylerc** [comment]
@yesbox I have been _super_ busy with a new job.  Really sorry for the delays.

I was previously unable to reproduce the bug, but I'll take another stab at it this week and see what happens.

**Comment by amarti2038** [comment]
Hi,

we also found the same bug. If single_repo = true we triggered the "ArgumentError at /node/version", after disabling it it works fine.

Regards,
Albert M.

**Comment by supertylerc** [comment]
Am I ever behind.

So what I've found is that the node isn't found and if it's not found we don't handle it gracefully.  It's detecting the group and the node name properly, but the method it's calling to find the node's version (`nodes.version`) can't find the node.  I'll dig into that method and see if I can find out why.

**Comment by supertylerc** [comment]
Looks like the reason this happens is because it's trying to find a repo named after the group anyway.  This is a bug in oxidized itself, not oxidized-web.

**Comment by supertylerc** [comment]
Good news: I know how to fix this specific bug!  But then there's another one.  :)  I'm following the rabbit.

**Comment by supertylerc** [comment]
More progress, but still more errors.  It's clear at this point that the `single_repo` implementation in core Oxidized has been severely broken for some time.

**Comment by supertylerc** [comment]
lol, I'm stopping now.  Right when I fixed everything and started to get ready to submit the PR, I discovered that the fix already exists in Oxidized, it just hasn't been published yet.

So...if you want this to be fixed, you need to install `oxidized` (not `oxidized-web`) from GitHub instead of RubyGems.

**Comment by supertylerc** [comment]
And, just to point to the actual commit that fixed this: https://github.com/ytti/oxidized/commit/60cd19857c2f722a1b39b8ee0f5b09440612d0ed

**Comment by amarti2038** [comment]
lol, ok thanks for your effort :)

**Comment by supertylerc** [comment]
Fix is now available in Oxidized 0.14.0, released a little bit ago.
